elitzur-core/src/main/scala/com/spotify/elitzur/validators/Validator.scala

```diff
@@ -244,6 +247,95 @@ object Validator extends Serializable {
     override def shouldValidate: Boolean = true
   }
 
+  //scalastyle:off method.length cyclomatic.complexity
+  /**
+   * Core validation loop for both dynamic and derived validators
+   * This code is deliberately mutable, uses casts to avoid unboxing, and avoids dereferencing
+   * Ignore Validators in order to optimize speed at runtime
+   */
+  @SuppressWarnings(Array("org.wartremover.warts.Var"))
+  def validationLoop[T](validatorAccessors: Array[ValidatorAccessor[Any]],
+                     constructor: Seq[Any] => T,
+                     outermostClassName: String,
+                     path: String = "",
+                     config: ValidationRecordConfig = DefaultRecordConfig)
+                    (implicit reporter: MetricsReporter): PostValidation[T] = {
+    val cs = new Array[Any](validatorAccessors.length)
+    var i = 0
+    var atLeastOneValid = false
+    var atLeastOneInvalid = false
+    val counterClassName = outermostClassName
```

Comment:
nit: it's not coming from this PR but we should probably drop the `counter` in counterClassName now that we've abstracted out counters. can probably drop this variable altogether since this is getting set a level higher now

Reply:
Good point, will fix!